src/SqlAsyncCollector.cs

```diff
@@ -151,7 +151,7 @@ private async Task UpsertRowsAsync(IEnumerable<T> rows, SqlAttribute attribute,
 
             int batchSize = 1000;
             await connection.OpenAsync();
```

Comment:
All of the lines after this before the try should go into the try so that we're always sure to close the connection once it's been opened. Currently if an exception occurs while doing the other stuff (unlikely but you never know) then we'll still never close the connection.